### PR TITLE
Add local resources support for oci-images

### PR DESCRIPTION
**Pull request by DomFleischmann**
This commit enables support for local resources of the oci-image type for charms.

### CODE REVIEW

File: tests/integration/oci-image-charm/dispatch [hunk 1]
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
+
+status-set active
```
[the remark this answers is not in the document]
Comment (johnsca) [reply tests/integration/oci-image-charm/dispatch:3]: I'm not certain that it wouldn't be better to have the `waiting` status be `blocked` instead. On the one hand, I don't want this to get stuck in `waiting`, but on the other, I feel (without any real evidence) like there's a chance that early hooks might have the controller fail on the `resource-get` even if the resource is provided with the deploy call and then have it succeed on a later hook.

File: tests/integration/oci-image-charm/metadata.yaml [hunk 2]
```diff
@@ -0,0 +1,10 @@
+name: oci-image-charm
+series: ["focal"]
+summary: "test"
+description: "test"
+maintainers: ["test"]
+resources:
+  oci-image:
+    type: oci-image
+    description: 'Backing OCI image'
+    auto-fetch: true
```
Comment (johnsca) [inline tests/integration/oci-image-charm/metadata.yaml:10]: This doesn't do anything and should be dropped. (It is used by a helper library in reactive charms, but that's not applicable here.)

**Review by johnsca — CHANGES_REQUESTED** [review]
There are some fixes that need to be made in the test, but the code itself LGTM.

**Review by johnsca — APPROVED** [review]

**Comment by johnsca** [comment]
!!build!!

File: tests/integration/test_model.py [hunk 3]
```diff
@@ -568,6 +568,27 @@ async def test_store_resources_charm(event_loop):
         assert ghost.units[0].workload_status == 'active'
 
 
+@base.bootstrapped
+@pytest.mark.asyncio
+async def test_local_oci_image_resource_charm(event_loop):
+    pytest.skip('test_local_resources_charm intermittent test failure')
+
+    tests_dir = Path(__file__).absolute().parent.parent
```
Comment (johnsca) [inline tests/integration/test_model.py:576]:
```suggestion
    tests_dir = Path(__file__).absolute().parent
```

**Review by johnsca — CHANGES_REQUESTED** [review]

**Review by johnsca — APPROVED** [review]

**Comment by johnsca** [comment]
Integration tests are still failing due to the snapd profile issue, not relevant to this change.  I confirmed locally that the added test works.

**Review by SimonRichardson — APPROVED** [review]

**Comment by SimonRichardson** [comment]
$$merge$$

**Comment by SimonRichardson** [comment]
!!build!!

**Comment by SimonRichardson** [comment]
Just fighting with CI atm, sorry for the noise.

**Comment by SimonRichardson** [comment]
!! build!!

**Comment by SimonRichardson** [comment]
!!build!!

**Comment by SimonRichardson** [comment]
$$merge$$

**Pull request merged**